crates/ruff_formatter/src/format_extensions.rs

```diff
@@ -1,178 +0,0 @@
-use crate::prelude::*;
-use std::cell::RefCell;
-use std::marker::PhantomData;
-use std::ops::Deref;
-
-use crate::Buffer;
-
-/// Utility trait that allows memorizing the output of a [Format].
-/// Useful to avoid re-formatting the same object twice.
-pub trait MemoizeFormat<Context> {
-    /// Returns a formattable object that memoizes the result of `Format` by cloning.
-    /// Mainly useful if the same sub-tree can appear twice in the formatted output because it's
-    /// used inside of `if_group_breaks` or `if_group_fits_single_line`.
-    ///
-    /// ```
-    /// use std::cell::Cell;
-    /// use ruff_formatter::{format, write};
-    /// use ruff_formatter::prelude::*;
-    /// use ruff_rowan::TextSize;
-    ///
-    /// struct MyFormat {
-    ///   value: Cell<u64>
-    /// }
-    ///
-    /// impl MyFormat {
-    ///     pub fn new() -> Self {
-    ///         Self { value: Cell::new(1) }
-    ///     }
-    /// }
-    ///
-    /// impl Format<SimpleFormatContext> for MyFormat {
-    ///     fn fmt(&self, f: &mut Formatter<SimpleFormatContext>) -> FormatResult<()> {
-    ///         let value = self.value.get();
-    ///         self.value.set(value + 1);
-    ///
-    ///         write!(f, [dynamic_text(&std::format!("Formatted {value} times."), TextSize::from(0))])
-    ///     }
-    /// }
-    ///
-    /// # fn main() -> FormatResult<()> {
-    /// let normal = MyFormat::new();
-    ///
-    /// // Calls `format` for everytime the object gets formatted
-    /// assert_eq!(
-    ///     "Formatted 1 times. Formatted 2 times.",
-    ///     format!(SimpleFormatContext::default(), [normal, space(), normal])?.print()?.as_code()
-    /// );
-    ///
-    /// // Memoized memoizes the result and calls `format` only once.
-    /// let memoized = normal.memoized();
-    /// assert_eq!(
-    ///     "Formatted 3 times. Formatted 3 times.",
-    ///     format![SimpleFormatContext::default(), [memoized, space(), memoized]]?.print()?.as_code()
-    /// );
-    /// # Ok(())
-    /// # }
-    /// ```
-    ///
-    fn memoized(self) -> Memoized<Self, Context>
-    where
-        Self: Sized + Format<Context>,
-    {
-        Memoized::new(self)
-    }
-}
-
-impl<T, Context> MemoizeFormat<Context> for T where T: Format<Context> {}
-
-/// Memoizes the output of its inner [Format] to avoid re-formatting a potential expensive object.
-#[derive(Debug)]
-pub struct Memoized<F, Context> {
```

Comment:
You probably want to keep this around. Memoization is essential to get good formatting performance when using the `BestFitting` IR